# Assignment: Build a Regression Model in Keras
## <a href="#parta">click here</a>

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

# Part (B)-- <a href="#partb">click here</a>

B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

# Part (D)-- <a href="#partd"> click here </a>

D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

In [5]:
# @title Importing Library
import pandas as pd
import numpy as np

# Library for modal
import keras
from keras.models import Sequential
from keras.layers import Dense

#For data spliting
from sklearn.model_selection import train_test_split

#For mean sqare error
from sklearn.metrics import mean_squared_error

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [6]:
# @title Loading data set-

concrete_data=pd.read_csv('https://cocl.us/concrete_data')

In [7]:
concrete_data.sample(5)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
446   165.0               128.5    132.1  175.1               8.1   
63    190.0               190.0      0.0  228.0               0.0   
977   313.3               145.0      0.0  178.5               8.0   
217   190.3                 0.0    125.2  161.9               9.9   
459   165.0               128.5    132.1  175.1               8.1   

     Coarse Aggregate  Fine Aggregate  Age  Strength  
446            1005.8           746.6   56     53.72  
63              932.0           670.0  270     50.66  
977            1001.9           688.7   28     41.05  
217            1088.1           802.6   56     38.56  
459            1005.8           746.6  100     55.02

**The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:**

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

### Let's check how many data points we have.

In [8]:
concrete_data.shape

(1030, 9)

In [9]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [10]:
concrete_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


In [11]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There are **1030** samples in the dataset.

**Strength is our target**

**The data looks very clean and is ready to be used to build our model.**

In [12]:
# Split data into predictors and target

predictors = concrete_data.iloc[:,:-1] # strength is the last column so this will exclude the last column.
target = concrete_data['Strength'] # Strength column

In [13]:
predictors.sample(5)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
943   151.6                 0.0    111.9  184.4               7.9   
489   387.0                20.0     94.0  157.0              13.9   
247   238.1                 0.0     94.1  186.7               7.0   
13    190.0               190.0      0.0  228.0               0.0   
125   531.3                 0.0      0.0  141.8              28.2   

     Coarse Aggregate  Fine Aggregate  Age  
943             992.0           815.9   28  
489             938.0           845.0    3  
247             949.9           847.0   56  
13              932.0           670.0   90  
125             852.1           893.7   28

In [14]:
target.head(3)

0    79.99
1    61.89
2    40.27
Name: Strength, dtype: float64

In [15]:
# No. of features
n_cols=predictors.shape[1]
n_cols

8

The below function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Function is using keras Sequantial that we have imported above

# <a id='parta'></a>

# Part-A

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) # hidden layers with node 10 and relu-- activation function
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [14]:
# Let's split the data in training and testing

X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

### Train and Test the Network

**Let's call the function now to create our model.**

In [ ]:
# build the model
# model = regression_model()
# epochs=50

# # Fit
# model.fit(X_train, y_train, epochs=epochs, verbose=1)

In [16]:
# Evaluate the model on the test data.

loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 327us/step


204.88458617373962

### Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [23]:
# Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.
# Making it as a function as we will use it later--
def iterate(epochs,total_mean_squared_errors):
#     total_mean_squared_errors = 50
#     epochs = 50
    mean_squared_errors = []
    
    model = regression_model()
    
    # Iterating 50 times
    for i in range(0, total_mean_squared_errors):
        X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
        model.fit(X_train, y_train, epochs=epochs, verbose=0)
        MSE = model.evaluate(X_test, y_test, verbose=0)
        print("MSE "+str(i+1)+": "+str(MSE))
        y_pred = model.predict(X_test)
        mean_square_error = mean_squared_error(y_test, y_pred)
        mean_squared_errors.append(mean_square_error)

    mean_squared_errors = np.array(mean_squared_errors)
    mean = np.mean(mean_squared_errors)
    standard_deviation = np.std(mean_squared_errors)

    print('\n')
    print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
    print("Mean: "+str(mean))
    print("Standard Deviation: "+str(standard_deviation))
    return (mean, standard_deviation)

In [24]:
mean_a, std_a = iterate(50,50)

MSE 1: 41.821701136221776
MSE 2: 43.209447101481906
MSE 3: 35.990279213124495
MSE 4: 49.58624949038607
MSE 5: 42.85005512669634
MSE 6: 40.467437312055175
MSE 7: 51.039800921690116
MSE 8: 35.54502662485857
MSE 9: 38.540394495991826
MSE 10: 38.60325090476224
MSE 11: 41.00113922415428
MSE 12: 39.125476479144545
MSE 13: 44.71271902534954
MSE 14: 45.81682416304801
MSE 15: 40.09976295051451
MSE 16: 32.443030940676195
MSE 17: 40.03720555722135
MSE 18: 41.327044237007215
MSE 19: 38.822374658677184
MSE 20: 38.407508769853216
MSE 21: 38.62789117016838
MSE 22: 41.19550398792649
MSE 23: 44.44575454810676
MSE 24: 43.96776868456004
MSE 25: 40.95668524214365
MSE 26: 43.64356542790978
MSE 27: 41.945716265335825
MSE 28: 35.1771556520925
MSE 29: 44.15007015339379
MSE 30: 39.31885284127541
MSE 31: 37.27677719955691
MSE 32: 33.335848157074075
MSE 33: 35.77686919672204
MSE 34: 52.59152011809611
MSE 35: 38.87970533648741
MSE 36: 43.811368911397494
MSE 37: 40.87903212883712
MSE 38: 41.57479754697929
MSE 39: 

<div id='partb'></div>

# Part-B

Result from previous part- 

The mean and standard deviation of 50 mean squared errors without normalized data. Total number of epochs for each training is: 50

Mean: 51.87652089428187
Standard Deviation: 8.24017812052958


**Here first job is normalisation.**

**--Then split data in train and test**

**--Build the modal**

**--Iterate for MSE and show mean and deviation**

In [30]:
#Normalising

predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [25]:
# Let's split the data in training and testing

X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

In [26]:
mean_b, std_b = iterate(epochs=50,total_mean_squared_errors=50)

MSE 1: 83.48188980497589
MSE 2: 81.62353009467758
MSE 3: 54.05317795392379
MSE 4: 50.130375155353235
MSE 5: 47.23215613010246
MSE 6: 49.45512101256732
MSE 7: 62.59146710738395
MSE 8: 44.11372364994777
MSE 9: 42.33838298019854
MSE 10: 47.77740364938878
MSE 11: 46.772023481844315
MSE 12: 37.323362109730546
MSE 13: 46.919747682836835
MSE 14: 52.5940397688486
MSE 15: 38.904310640007935
MSE 16: 33.64142221308835
MSE 17: 40.094195899839924
MSE 18: 46.40260771717454
MSE 19: 36.056612835732864
MSE 20: 37.528404371638125
MSE 21: 41.56348250058863
MSE 22: 36.91549457932753
MSE 23: 36.07111458639497
MSE 24: 35.16788892529929
MSE 25: 53.95384874313009
MSE 26: 38.96927208267755
MSE 27: 38.903917294104126
MSE 28: 34.39932554670908
MSE 29: 41.76968213423942
MSE 30: 38.86033421735547
MSE 31: 42.00711006485528
MSE 32: 31.74198700000553
MSE 33: 34.486219906112524
MSE 34: 40.21352933371337
MSE 35: 39.23569826786572
MSE 36: 39.550515998914406
MSE 37: 41.11803272086826
MSE 38: 41.49933239094262
MSE 39: 37.

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/keras/callbacks.py:120: UserWarning: Method on_batch_end() is slow compared to the batch update (0.100737). Check your callbacks.
  % delta_t_median)


MSE 47: 41.84637574624861
MSE 48: 40.134507015685045
MSE 49: 41.09470255089423
MSE 50: 39.38209086791597


Below is the mean and standard deviation of 50 mean squared errors without normalized data. Total number of epochs for each training is: 50

Mean: 43.00947793675874
Standard Deviation: 10.134744586229775


**Here now we will see the change in mean and standard deviation**

In [3]:
# change=[(mean_a, std_a),(mean_b, std_b)]
# pd.DataFrame(change,index=['A','B'],columns=['Mean','STD'])

<div id='partd'>

# Part-D

D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.


In [27]:
# define a new- regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))# hidden layers with node 10 and relu-- activation function
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [ ]:
mean_d, std_d = iterate(epochs=50,total_mean_squared_errors=50)

MSE 1: 93.53098257614185
MSE 2: 81.11631888479091
MSE 3: 55.667201872396625
MSE 4: 55.12707702241669
MSE 5: 47.29528346879582
MSE 6: 46.35549650223124
MSE 7: 55.74195296401731
MSE 8: 41.74331145919257
MSE 9: 42.91709949283538
MSE 10: 41.43547990175512
MSE 11: 40.88307713999332
MSE 12: 37.79117933600466
MSE 13: 50.418074647971345
MSE 14: 42.56416778811359
MSE 15: 38.01758824814485
MSE 16: 32.65989997086016
MSE 17: 39.06480032417767
MSE 18: 42.694489685848694
MSE 19: 33.85735749426783
MSE 20: 39.63545914683913
MSE 21: 37.94860595406838
MSE 22: 40.662255518644756
MSE 23: 32.395512170390404
MSE 24: 36.137561933893984
MSE 25: 38.806665883480925
MSE 26: 40.612192962547724
MSE 27: 37.33787232926748
MSE 28: 35.17908586421831
MSE 29: 40.175044717140565
MSE 30: 41.25339290779385
MSE 31: 37.59982294866568


In [ ]:
# Here comes the table for difference
change=[(mean_b, std_b),(mean_d, std_d),(mean_d-mean_b, std_d-std_b)]
pd.DataFrame(change,index=['B','D','Change'],columns=['Mean','STD'])